openfisca_core/formulas.py

```diff
@@ -834,110 +823,71 @@ def new_filled_column(
+    # Turn function into a dated function
+    formula_name_prefix = 'formula'
+    formula_name_separator = '_'
+    dated_formulas_class = []
+    for function_name, function in specific_attributes.copy().iteritems():
+        if not function_name.startswith(formula_name_prefix):
+            # Current attribute isn't a formula
+            continue
+
+        formula_start_date = extract_formula_date_from_name(function_name, formula_name_prefix, formula_name_separator)
+        if end is not None:
+            end_date = get_datetime_date(name, end)
+            assert end_date >= formula_start_date, \
+                'You declared that "{}" ends on "{}", but you wrote a formula to calculate it from "{}" ({}). The "end" attribute of a variable must be posterior to the start dates of all its formulas.'.format(name, end_date, formula_start_date, function_name)
+        dated_formula_class_attributes = formula_class_attributes.copy()
+        dated_formula_class_attributes['formula'] = function
+        dated_formula_class = type(name.encode('utf-8'), (Formula,), dated_formula_class_attributes)
+
+        del specific_attributes[function_name]
+        dated_formulas_class.append(dict(
+            formula_class = dated_formula_class,
+            start_instant = periods.instant(formula_start_date),
+            ))
 
-        formula_class_attributes['dated_formulas_class'] = dated_formulas_class
-    else:
-        assert issubclass(formula_class, SimpleFormula), formula_class
+    # Sort dated formulas by start instant.
+    dated_formulas_class.sort(key = lambda dated_formula_class: dated_formula_class['start_instant'])
+
+    # Add dated formulas defined in (optional) reference column when they are not overridden by new dated
+    # formulas.
+    if reference_column is not None and issubclass(reference_column.formula_class, Formula):
+        for reference_dated_formula_class in reference_column.formula_class.dated_formulas_class:
+            reference_dated_formula_class = reference_dated_formula_class.copy()
+            for dated_formula_class in dated_formulas_class:
+                if reference_dated_formula_class['start_instant'] >= dated_formula_class['start_instant']:
+                    break
+
+            else:
+                dated_formulas_class.append(reference_dated_formula_class)
+        dated_formulas_class.sort(key = lambda dated_formula_class: dated_formula_class['start_instant'])
 
-        function = specific_attributes.pop('function', None)
-        if column.definition_period == ETERNITY:
-            assert function is None
-        if reference_column is not None and function is None:
-            function = reference_column.formula_class.function
-        formula_class_attributes['function'] = function
+    formula_class_attributes['dated_formulas_class'] = dated_formulas_class
 
     # Ensure that all attributes defined in ConversionColumn class are used.
```

Comment:
Nothing to do with your PR, but this line is outdated (`ConversionColumn` do not exist), we can remove it.